pylsp_ruff/ruff_lint.py

```diff
@@ -0,0 +1,258 @@
+import json
+import logging
+import sys
+from pathlib import PurePath
+from subprocess import PIPE, Popen, SubprocessError
+
+from pylsp import hookimpl, lsp
+from pylsp._utils import find_parents
+from pylsp.workspace import Document, Workspace
+
+# Use built-in tomllib for python>=3.11
+if sys.version_info >= (3, 11):
+    try:
+        import tomllib
+    except ImportError:
+        import tomli as tomllib
+else:
+    import tomli as tomllib
+
+log = logging.getLogger(__name__)
+
+UNNECESSITY_CODES = {
+    "F401",  # `module` imported but unused
+    "F504",  # % format unused named arguments
+    "F522",  # .format(...) unused named arguments
+    "F523",  # .format(...) unused positional arguments
+    "F841",  # local variable `name` is assigned to but never used
+}
+
+
+@hookimpl
+def pylsp_settings():
+    # flake8 and pycodestyle disabled by default with this plugin
+    return {
+        "plugins": {
+            "ruff": {
+                "enabled": True,
+                "config": None,
+                "exclude": None,
+                "executable": "ruff",
+                "ignore": None,
+                "lineLength": None,
+                "perFileIgnores": None,
+                "select": None,
+            },
+            "flake8": {"enabled": False},
+            "pycodestyle": {"enabled": False},
+        }
+    }
+
+
+@hookimpl
+def pylsp_lint(workspace: Workspace, document: Document) -> list:
+    """Register ruff as the linter
+
+    Parameters
+    ----------
+    workspace : pylsp.workspace.Workspace
+        Current workspace.
+    document : pylsp.workspace.Document
+        Document to apply ruff on.
+
+    Returns
+    -------
+    list of dicts containing the diagnostics.
+    """
+    config = workspace._config
+    settings = config.plugin_settings("ruff", document_path=document.path)
+    log.debug(f"Got ruff settings: {settings}")
+
+    args_dict = load_config(workspace, document)
+    ruff_executable = args_dict.pop("executable")
+    args = build_args(document, args_dict)
+
+    output = run_ruff_lint(ruff_executable, document, args)
+    return parse_ruff_stdout(output)
+
+
+def run_ruff_lint(ruff_executable: str, document: Document, arguments: list) -> str:
+    """Run ruff on the given document and the given arguments.
+
+    Parameters
+    ----------
+        ruff_executable : str
+            Path to the executable.
+        document : pylsp.workspace.Document
+            File to run ruff on.
+        arguments : list
+            Arguments to provide for ruff.
+
+    Returns
+    -------
+        str containing the result in json format.
+    """
+    log.debug(f"Calling {ruff_executable} with args: {arguments} on '{document.path}'")
+    try:
+        cmd = [ruff_executable]
+        cmd.extend(arguments)
```

Comment:
Could you try something like:

```py
cmd = ["ruff_executable"]
if arguments:
  cmd.append("--")
  cmd.extend(arguments)
```

Reply:
I'm not sure if I follow, that would try to run `ruff -- - --quiet --format=json --line-length=88` which results in an error.
Can explain the '--'?

Reply:
Forgive me for butting in, but I'm keen to integrate Ruff into my LSP setup. :slightly_smiling_face: 

`--` in shell commands signifies the end of options. It prevents CLI programs from attempting to parse positional arguments starting with `-` as options.

A common example would be attempting to grep a file for the string `-E`. If you tried `grep -E file`, then grep would think you passed it the `-E` option. To prevent it from doing that, you could run `grep -- -E file`. grep would then know to search `file` for the string `-E`.

In this case, I expect you'd want to ensure your call to ruff *ends* with `-- -` to remove any ambiguity in the command. Adapting your example, you'd probably want to end up with `ruff --quiet --format=json --line-length=88 -- -`.

Reply:
Sorry, I lost track of this one -- thanks @smjreynolds.

Reply:
I see, thanks for clearing that up, that's a good idea to add :+1: